flow/workflows/maintenance_flow.go

```diff
@@ -0,0 +1,275 @@
+	for range mirrorsList.Mirrors {
+		selector.Select(ctx)
+		if err := ctx.Err(); err != nil {
+			return activities.MaintenanceMirrorsInfo{}, err
+		}
+	}
+	for i, mirror := range mirrorsList.Mirrors {
+		if runningMirrors[i] {
+			onlyRunningMirrors = append(onlyRunningMirrors, mirror)
+		}
+	}
+	return activities.MaintenanceMirrorsInfo{
+		Mirrors: onlyRunningMirrors,
+	}, nil
+}
+
+func getAllMirrors(ctx workflow.Context) (activities.MaintenanceMirrorsInfo, error) {
+	getMirrorsFuture := workflow.ExecuteActivity(ctx, maintenance.GetAllMirrors)
+	var mirrorsList activities.MaintenanceMirrorsInfo
+	err := getMirrorsFuture.Get(ctx, &mirrorsList)
+	return mirrorsList, err
+}
+
+func EndMaintenanceWorkflow(ctx workflow.Context, input *protos.EndMaintenanceFlowInput) (*protos.EndMaintenanceFlowOutput, error) {
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Starting EndMaintenance workflow", "input", input)
+	defer runBackgroundAlerter(ctx)()
+
+	flowOutput, err := endMaintenance(ctx, logger)
+	if err != nil {
+		slog.Error("Error in EndMaintenance workflow", "error", err)
+		return nil, err
+	}
+	return flowOutput, nil
+}
+
+func endMaintenance(ctx workflow.Context, logger log.Logger) (*protos.EndMaintenanceFlowOutput, error) {
+	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: 24 * time.Hour,
+		HeartbeatTimeout:    1 * time.Minute,
+	})
+
+	mirrorsList, err := resumeBackedUpMirrors(ctx, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	clearBackupsFuture := workflow.ExecuteActivity(ctx, maintenance.CleanupBackupedFlows)
+	err = clearBackupsFuture.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("Resumed backed up mirrors", "mirrors", mirrorsList)
+
+	future := workflow.ExecuteActivity(ctx, maintenance.DisableMaintenanceMode)
+	err = future.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("EndMaintenance workflow completed")
+	return &protos.EndMaintenanceFlowOutput{
+		Version: peerdbenv.PeerDBVersionShaShort(),
+	}, nil
+}
+
+func resumeBackedUpMirrors(ctx workflow.Context, logger log.Logger) (activities.MaintenanceMirrorsInfo, error) {
+	future := workflow.ExecuteActivity(ctx, maintenance.GetBackedUpFlows)
+	var mirrorsList activities.MaintenanceMirrorsInfo
+	err := future.Get(ctx, &mirrorsList)
+	if err != nil {
+		return activities.MaintenanceMirrorsInfo{}, err
+	}
+
+	selector := workflow.NewSelector(ctx)
+	for _, mirror := range mirrorsList.Mirrors {
+		activityInput := mirror
+		f := workflow.ExecuteActivity(
+			ctx,
+			maintenance.ResumeMirror,
+			activityInput,
+		)
+
+		selector.AddFuture(f, func(f workflow.Future) {
+			err := f.Get(ctx, nil)
+			if err != nil {
+				logger.Error("Error resuming mirror", "mirror", mirror, "error", err)
+			} else {
+				logger.Info("Finished resuming mirror", "mirror", mirror)
+			}
+			ctx.Done()
+		})
+	}
+
+	for range mirrorsList.Mirrors {
+		selector.Select(ctx)
```

Comment:
Fun quirk: `selector.Select(ctx)` only uses ctx for metadata. If ctx completes while Select running it won't return. You have `selector.AddReceive(ctx.Done(), ...)` to not ignore workflow being cancelled while selecting here. See other workflows for examples of that

Tho in this case context cancel will complete the activies which should complete these futures